lib/Utils/Layout/Hoisting.cpp

```diff
@@ -0,0 +1,119 @@
+#include "lib/Utils/Layout/Hoisting.h"
+
+#include "lib/Utils/Layout/IslConversion.h"
+#include "lib/Utils/Layout/Utils.h"
+#include "llvm/include/llvm/ADT/SmallVector.h"  // from @llvm-project
+#include "mlir/include/mlir/Analysis/Presburger/IntegerRelation.h"  // from @llvm-project
+
+// ISL
+#include "include/isl/ast.h"             // from @isl
+#include "include/isl/ast_build.h"       // from @isl
+#include "include/isl/ast_type.h"        // from @isl
+#include "include/isl/constraint.h"      // from @isl
+#include "include/isl/ctx.h"             // from @isl
+#include "include/isl/local_space.h"     // from @isl
+#include "include/isl/map.h"             // from @isl
+#include "include/isl/map_type.h"        // from @isl
+#include "include/isl/set.h"             // from @isl
+#include "include/isl/space.h"           // from @isl
+#include "include/isl/space_type.h"      // from @isl
+#include "include/isl/union_map.h"       // from @isl
+#include "include/isl/union_map_type.h"  // from @isl
+
+namespace mlir {
+namespace heir {
+
+using llvm::SmallVector;
+using presburger::BoundType;
+using presburger::IntegerRelation;
+using presburger::PresburgerSpace;
+using presburger::VarKind;
+
+presburger::IntegerRelation hoistConversionThroughMatvec(
+    const IntegerRelation& matrixLayout, const IntegerRelation& fromVecLayout,
+    const IntegerRelation& toVecLayout) {
+  // The intuition for this function is that the conversion from fromVecLayout
+  // to toVecLayout implies some transformation of the slot ordering of the
+  // packed vector. The kernels we support have matrix layouts for which the
+  // packed slots of a ciphertext track the columns of the packed vector.
+  // So we need to apply the same transformation of the packed vector slots
+  // to the packed matrix slots.
+  //
+  // This function works in two steps:
+  //
+  // 1. Compute the inferred re-packing relation of vector slots (i.e., a
+  // relation (ct, slots) -> (ct, slots))
+  //
+  // 2. Compose (1) with the (ct, slot) dims of the matrix packing.
+
+  IntegerRelation fromClone(fromVecLayout);
+  IntegerRelation toClone(toVecLayout);
+
+  // Project out the ciphertext dimension, though this will need to change when
+  // we get a larger vector than can fit in one ciphertext. This is where the
+  // assumption about the vector packing fitting into one ciphertext comes from.
+  fromClone.projectOut(1);
+  toClone.projectOut(1);
+
+  fromClone.inverse();
+  fromClone.compose(toClone);
+  fromClone.removeRedundantConstraints();
+  fromClone.simplify();
+
+  // Put the ct dim back in with bounds from the matrix layout, and equality
+  // between the domain ct var and the range ct var.
+  fromClone.insertVar(VarKind::Domain, 0, 1);
+  fromClone.insertVar(VarKind::Range, 0, 1);
+  std::optional<int64_t> ctUb =
+      matrixLayout.getConstantBound64(BoundType::UB, 2);
```

Comment:
```suggestion
      matrixLayout.getConstantBound64(BoundType::UB, matrixLayout.getVarKindOffset(VarKind:Range);
```

should we just make helpers as part of the tensor_ext dialect like `getCtOffset(relation)` and `getSlotOffset(relation)`?

Reply:
Done. I don't think the helpers are needed right now, but maybe in the future it would make sense to do this at a higher level (e.g., a wrapper type for a CiphertextPacking around IntegerRelation that could expose a method like getNumCiphertexts)